### PR TITLE
Break sentence scope on [.!?]

### PR DESCRIPTION
**Pull request by AndreasArvidsson**
sbd would sometimes not break lines ending with [.!?] when `new_line_boundaries: false` is used

## Checklist

- [x] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [/] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet)
- [/] I have not broken the cheatsheet

### CODE REVIEW

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/SentenceScopeHandler/SentenceSegmenter.ts [hunk 1]
```diff
@@ -4,7 +4,7 @@ import { MatchedText, matchRegex, testRegex } from "../../../../util/regex";
 // A sentence starts with a letter
 const leadingOffsetRegex = /\p{L}/u;
 // A line with no letters is invalid and breaks sentences
-const invalidLineRegex = /(\n[^\p{L}]*\n)/gu;
+const invalidLineRegex = /(\n[^\p{L}]*\n)|(?<=[.!?])(\s*\r?\n)/gu;
```
Comment (pokey) [inline packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/SentenceScopeHandler/SentenceSegmenter.ts:7]: this regex confuses me a bit:

- the first part is matching an entire line including newlines but the second part is not
- the second part is handling `\r` but the first part never mentions it

I'm not saying it's necessarily incorrect, but I don't understand it.  Would it be possible to fix and/or update the comment?

File: packages/cursorless-vscode-e2e/src/suite/fixtures/recorded/scopes/changeEverySentence.yml
Comment (pokey) [inline packages/cursorless-vscode-e2e/src/suite/fixtures/recorded/scopes/changeEverySentence.yml:1]: this test covers everything that the other two sentence tests do, no? I'd be tempted to remove those

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/SentenceScopeHandler/SentenceSegmenter.ts (hunk 1 above)
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/SentenceScopeHandler/SentenceSegmenter.ts]: better?

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/SentenceScopeHandler/SentenceSegmenter.ts [hunk 2]
```diff
@@ -1,10 +1,11 @@
...
 import { MatchedText, matchRegex, testRegex } from "../../../../util/regex";
 
-// A sentence starts with a letter
-const leadingOffsetRegex = /\p{L}/u;
+// A sentence starts with a letter with adjacent leading symbols. Whitespace excluded.
+const leadingOffsetRegex = /\S*\p{L}/u;
 // A line with no letters is invalid and breaks sentences
-const invalidLineRegex = /(\n[^\p{L}]*\n)/gu;
+// Also break lines ending with [.!?]
+const invalidLineRegex = /(\r?\n[^\p{L}]*\r?\n)|(?<=[.!?])(\s*\r?\n)/gu;
```
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/SentenceScopeHandler/SentenceSegmenter.ts]: probably not. Updated

File: packages/cursorless-vscode-e2e/src/suite/fixtures/recorded/scopes/sentence/changeEverySentence.yml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    target:
+      type: primitive
+      modifiers:
+        - type: everyScope
+          scopeType: {type: sentence}
+  usePrePhraseSnapshot: true
+initialState:
+  documentContents: |-
+    1. **Custom subtitles** - User customizable subtitles for Talon. 
+        - [Subtitles and notifications](./core/on_phrase/subtitles_and_notifications)
```
Comment (pokey) [inline packages/cursorless-vscode-e2e/src/suite/fixtures/recorded/scopes/sentence/changeEverySentence.yml:16]: so were both of these sentences broken or just the first one?

Comment (AndreasArvidsson) [reply packages/cursorless-vscode-e2e/src/suite/fixtures/recorded/scopes/sentence/changeEverySentence.yml:16]: It was just one sentence

**Comment by pokey** [comment]
Hmm looks like something strange is going on here.  Based on some experiments using [their web demo](https://tessmore.github.io/sbd/), it breaks the sentences correctly if `Talon` is lowercase 🤔.  I'm hesitant to just override their logic for this specific case, as it seems to be more complex than just their logic breaking whenever a line ends with `.`

**Comment by AndreasArvidsson** [comment]
Issue file with sbd and Cursorless. Issue added to fix me comment.
https://github.com/cursorless-dev/cursorless/issues/1753

File: packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/SentenceScopeHandler/SentenceSegmenter.ts [hunk 4]
```diff
@@ -1,10 +1,13 @@
...
 
-// A sentence starts with a letter
-const leadingOffsetRegex = /\p{L}/u;
+// A sentence starts with a letter with adjacent leading symbols. Whitespace excluded.
+const leadingOffsetRegex = /\S*\p{L}/u;
 // A line with no letters is invalid and breaks sentences
-const invalidLineRegex = /(\n[^\p{L}]*\n)/gu;
+// Also break lines ending with [.!?]
+// FIX ME: Remove second term once bug is fixed.
+// https://github.com/cursorless-dev/cursorless/issues/1753
```
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/SentenceScopeHandler/SentenceSegmenter.ts]: done

**Review by pokey — APPROVED** [review]
Looks good to me; made some minor tweaks.  Merge if happy

**Pull request merged**